### DISCUSSION TITLE
Yaw optimization serial refine - Ny passes

### DISCUSSION THREAD

**Discussion by fwasilcz** [NatLabRockies/floris#1013] [Q&A] [answered]
Hi,
I have a question about the meaning of second and further values in Ny passes matrix in the Serial Refine method for yaw optimization. Based on the paper on the method I'm guessing that the first term in Ny passes matrix is the number of angles that are evaluated between minimum_yaw_angle and maximum_yaw_angle. Following that logic, the next terms mean number of passes during the "Refine" step of the optimization. If this is the case, then how are the minimum and maximum angle set for this step? As a fraction of initial min and max, or some other way? Are the further terms another level of "Refine" for even smaller angle range? I'm trying to understand the method better.
Regards
Filip

**Comment by paulf81** [comment, marked as answer]
Hi yes, you have the idea right.  If you look to line 289 of `yaw_optimizer_sr.py` you'll see the relavent code for generating the points to test in the refine pass:

```
                # Update bounds for next iteration to close proximity of optimal solution
                dx = (
                    evaluation_grid[1, :, :] -
                    evaluation_grid[0, :, :]
                )[ids]
                self._yaw_lbs[ids] = np.clip(
                    yaw_angles_opt[ids] - 0.50 * dx,
                    self._minimum_yaw_angle_subset[ids],
                    self._maximum_yaw_angle_subset[ids]
                )
                self._yaw_ubs[ids] = np.clip(
                    yaw_angles_opt[ids] + 0.50 * dx,
                    self._minimum_yaw_angle_subset[ids],
                    self._maximum_yaw_angle_subset[ids]
                )

```

This code is updating the bounds so that during the second pass through `_generate_evaluation_grid` this line:

```
 yaw_angles_subset = np.linspace(yaw_lb, yaw_ub, Ny + 1)[ids]
```

Generates the test points in between the bounds, using the number of steps given by the Ny passes matrix

**Reply by fwasilcz** [reply]
Thank you for the clarification!
